src/ngraph/runtime/cpu/pass/cpu_memory_optimization.cpp

```diff
@@ -0,0 +1,188 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/runtime/cpu/pass/cpu_memory_optimization.hpp"
+
+#include "ngraph/descriptor/output.hpp"
+#include "ngraph/op/concat.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/runtime/cpu/cpu_op_annotations.hpp"
+
+using namespace ngraph;
+
+bool runtime::cpu::pass::CPUMemoryOptimization::run_on_function(std::shared_ptr<Function> function)
+{
+    for (auto n : function->get_ordered_ops())
+    {
+        if (auto concat = std::dynamic_pointer_cast<op::Concat>(n))
+        {
+            auto shape = concat->get_input_shape(0);
+            auto axis = concat->get_concatenation_axis();
+            auto product = 1;
+            for (int i = 0; i < axis; i++)
+            {
+                product *= shape[i];
+            }
+            if (product != 1)
+            {
+                NGRAPH_DEBUG << "cpu_post_layout_assignment: The product of Concat's shape "
+                                "before concat axis is not 1, no in place concat";
+                continue;
+            }
+
+            bool in_place_concat = true;
+
+            for (descriptor::Input& input : concat->get_inputs())
+            {
+                if (shape_size(input.get_shape()) == 0)
+                {
+                    NGRAPH_DEBUG << "cpu_post_layout_assignment: 0 length tensor, no in "
+                                    "place concat";
+                    in_place_concat = false;
+                    break;
```

Comment:
@jbobba we aren't running `ZeroDimElimination` on CPU 😉 fyi : @ayzhuang   since we don't have any issues handling those comparing to the other backends 😄